// vue3-tournament/dist/tournament-bracket.mjs
import { defineComponent as O, openBlock as C, createElementBlock as b, createElementVNode as u, renderSlot as l, normalizeProps as r, guardReactiveProps as c, toDisplayString as T, normalizeClass as f, createCommentVNode as p, computed as P, resolveComponent as v, createVNode as N, withCtx as d, ref as M, normalizeStyle as I } from "vue";
const S = O({
  name: "GameMatch",
  props: {
    bracketNode: {
      type: Object,
      default: () => ({})
    },
    highlightedTeamId: {
      type: [String, Number],
      default: void 0
    }
  },
  setup(e, { emit: a }) {
    return {
      getPlayerClass: (t) => {
        let n = "";
        return e.highlightedTeamId === t.id && !t.disabled && (n += " highlight"), t.disabled && (n += " disabled"), n;
      },
      onClick: (t) => {
        var n, o, h;
        a("onMatchClick", (o = (n = e.bracketNode) == null ? void 0 : n.match) == null ? void 0 : o.id), a("onParticipantClick", t, (h = e.bracketNode) == null ? void 0 : h.match);
      },
      highlightTeam: (t) => {
        a("onSelectedTeam", t);
      },
      unhighlightTeam: () => {
        a("onDeselectedTeam");
      }
    };
  }
}), y = (e, a) => {
  const s = e.__vccOpts || e;
  for (const [m, g] of a)
    s[m] = g;
  return s;
}, B = { class: "vt-item-feedIn" }, D = { class: "title" }, w = { class: "name" };
function z(e, a, s, m, g, k) {
  var t, n;
  return C(), b("div", B, [
    u("div", null, [
      l(e.$slots, "match.title", r(c({ match: (t = e.bracketNode) == null ? void 0 : t.match })), () => {
        var o, h;
        return [
          u("div", D, [
            u("span", null, T((h = (o = e.bracketNode) == null ? void 0 : o.match) == null ? void 0 : h.title), 1)
          ])
        ];
      }),
      u("div", {
        class: f(["vt-feedIn", e.getPlayerClass(e.bracketNode.match.feedIn)]),
        onMouseover: a[0] || (a[0] = (o) => e.highlightTeam(e.bracketNode.match.feedIn.id)),
        onMouseleave: a[1] || (a[1] = (...o) => e.unhighlightTeam && e.unhighlightTeam(...o)),
        onClick: a[2] || (a[2] = (o) => e.onClick(e.bracketNode.match.feedIn))
      }, [
        l(e.$slots, "feedIn", r(c({ team: (n = e.bracketNode.match) == null ? void 0 : n.feedIn, match: e.bracketNode.match })), () => {
          var o, h;
          return [
            u("span", w, T((h = (o = e.bracketNode.match) == null ? void 0 : o.feedIn) == null ? void 0 : h.name), 1)
          ];
        })
      ], 34)
    ])
  ]);
}
const G = /* @__PURE__ */ y(S, [["render", z]]), R = O({
  name: "GameMatch",
  props: {
    bracketNode: {
      type: Object,
      default: () => ({})
    },
    highlightedTeamId: {
      type: [String, Number],
      default: void 0
    }
  },
  setup(e, { emit: a }) {
    return {
      getPlayerClass: (t) => {
        var o, h;
        let n = "";
        return ((o = e.bracketNode.match) == null ? void 0 : o.winner) !== void 0 && t.id === ((h = e.bracketNode.match) == null ? void 0 : h.winner) && (n = "winner"), e.highlightedTeamId === t.id && !t.disabled && (n += " highlight"), t.disabled && (n += " disabled"), n;
      },
      onClick: (t) => {
        var n, o, h;
        a("onMatchClick", (o = (n = e.bracketNode) == null ? void 0 : n.match) == null ? void 0 : o.id), a("onParticipantClick", t, (h = e.bracketNode) == null ? void 0 : h.match);
      },
      highlightTeam: (t) => {
        a("onSelectedTeam", t);
      },
      unhighlightTeam: () => {
        a("onDeselectedTeam");
      }
    };
  }
}), V = { class: "vt-item-teams" }, j = { class: "title" }, A = { class: "name" }, F = {
  key: 0,
  class: "score"
}, E = { class: "name" }, q = {
  key: 0,
  class: "score"
};
function H(e, a, s, m, g, k) {
  return C(), b("div", V, [
    u("div", null, [
      l(e.$slots, "match.title", r(c({ match: e.bracketNode.match })), () => {
        var t;
        return [
          u("div", j, [
            u("span", null, T((t = e.bracketNode.match) == null ? void 0 : t.title), 1)
          ])
        ];
      }),
      u("div", {
        class: f([
          "vt-team",
          "vt-team-top",
          e.getPlayerClass(e.bracketNode.match.team1)
        ]),
        onMouseover: a[0] || (a[0] = (t) => e.highlightTeam(e.bracketNode.match.team1.id)),
        onMouseleave: a[1] || (a[1] = (...t) => e.unhighlightTeam && e.unhighlightTeam(...t)),
        onClick: a[2] || (a[2] = (t) => e.onClick(e.bracketNode.match.team1))
      }, [
        l(e.$slots, "team", r(c({ team: e.bracketNode.match.team1, match: e.bracketNode.match })), () => [
          l(e.$slots, "team.name", r(c({ team: e.bracketNode.match.team1, match: e.bracketNode.match })), () => [
            u("span", A, T(e.bracketNode.match.team1.name), 1)
          ]),
          l(e.$slots, "team.score", r(c({ team: e.bracketNode.match.team1, match: e.bracketNode.match })), () => [
            e.bracketNode.match.team1.score != null && e.bracketNode.match.team1.score >= 0 ? (C(), b("span", F, T(e.bracketNode.match.team1.score), 1)) : p("", !0)
          ])
        ])
      ], 34),
      u("div", {
        class: f([
          "vt-team",
          "vt-team-bot",
          e.getPlayerClass(e.bracketNode.match.team2)
        ]),
        onMouseover: a[3] || (a[3] = (t) => e.highlightTeam(e.bracketNode.match.team2.id)),
        onMouseleave: a[4] || (a[4] = (...t) => e.unhighlightTeam && e.unhighlightTeam(...t)),
        onClick: a[5] || (a[5] = (t) => e.onClick(e.bracketNode.match.team2))
      }, [
        l(e.$slots, "team", r(c({ team: e.bracketNode.match.team2, match: e.bracketNode.match })), () => [
          l(e.$slots, "team.name", r(c({ team: e.bracketNode.match.team2, match: e.bracketNode.match })), () => [
            u("span", E, T(e.bracketNode.match.team2.name), 1)
          ]),
          l(e.$slots, "team.score", r(c({ team: e.bracketNode.match.team2, match: e.bracketNode.match })), () => [
            e.bracketNode.match.team2.score != null && e.bracketNode.match.team2.score >= 0 ? (C(), b("span", q, T(e.bracketNode.match.team2.score), 1)) : p("", !0)
          ])
        ])
      ], 34)
    ])
  ]);
}
const J = /* @__PURE__ */ y(R, [["render", H]]), K = O({
  name: "BracketNode",
  components: {
    GameMatch: J,
    FeedIn: G
  },
  props: {
    bracketNode: {
      type: Object,
      default: () => ({})
    },
    highlightedTeamId: {
      type: [String, Number],
      default: void 0
    }
  },
  setup(e, { emit: a }) {
    return {
      matchArePresent: P(() => e.bracketNode.match !== void 0),
      getBracketNodeClass: (o) => o.children[0] || o.children[1] ? "vt-item-parent" : o.hasParent ? "vt-item-child" : "",
      onMatchClick: (o) => {
        a("onMatchClick", o);
      },
      highlightTeam: (o) => {
        a("onSelectedTeam", o);
      },
      unhighlightTeam: () => {
        a("onDeselectedTeam");
      },
      onParticipantClick: (o, h) => {
        a("onParticipantClick", o, h);
      }
    };
  }
}), L = {
  key: 0,
  class: "vt-item"
}, Q = {
  key: 2,
  class: "vt-item-children"
}, U = {
  key: 0,
  class: "vt-item-child"
}, W = {
  key: 1,
  class: "vt-item-child"
};
function X(e, a, s, m, g, k) {
  var h, $;
  const t = v("FeedIn"), n = v("GameMatch"), o = v("BracketNode", !0);
  return e.matchArePresent ? (C(), b("div", L, [
    (($ = (h = e.bracketNode) == null ? void 0 : h.match) == null ? void 0 : $.feedIn) !== void 0 ? (C(), b("div", {
      key: 0,
      class: f(e.getBracketNodeClass(e.bracketNode))
    }, [
      N(t, {
        "bracket-node": e.bracketNode,
        "highlighted-team-id": e.highlightedTeamId,
        onOnSelectedTeam: e.highlightTeam,
        onOnDeselectedTeam: e.unhighlightTeam,
        onOnMatchClick: e.onMatchClick,
        onOnParticipantClick: e.onParticipantClick
      }, {
        "match.title": d((i) => [
          l(e.$slots, "match.title", r(c(i)))
        ]),
        feedIn: d((i) => [
          l(e.$slots, "feedIn", r(c(i)))
        ]),
        _: 3
      }, 8, ["bracket-node", "highlighted-team-id", "onOnSelectedTeam", "onOnDeselectedTeam", "onOnMatchClick", "onOnParticipantClick"])
    ], 2)) : (C(), b("div", {
      key: 1,
      class: f(e.getBracketNodeClass(e.bracketNode))
    }, [
      N(n, {
        "bracket-node": e.bracketNode,
        "highlighted-team-id": e.highlightedTeamId,
        onOnSelectedTeam: e.highlightTeam,
        onOnDeselectedTeam: e.unhighlightTeam,
        onOnMatchClick: e.onMatchClick,
        onOnParticipantClick: e.onParticipantClick
      }, {
        "match.title": d((i) => [
          l(e.$slots, "match.title", r(c(i)))
        ]),
        team: d((i) => [
          l(e.$slots, "team", r(c(i)))
        ]),
        "team.name": d((i) => [
          l(e.$slots, "team.name", r(c(i)))
        ]),
        "team.score": d((i) => [
          l(e.$slots, "team.score", r(c(i)))
        ]),
        _: 3
      }, 8, ["bracket-node", "highlighted-team-id", "onOnSelectedTeam", "onOnDeselectedTeam", "onOnMatchClick", "onOnParticipantClick"])
    ], 2)),
    e.bracketNode.children[0] || e.bracketNode.children[1] ? (C(), b("div", Q, [
      e.bracketNode.children[0] ? (C(), b("div", U, [
        N(o, {
          "bracket-node": e.bracketNode.children[0],
          "highlighted-team-id": e.highlightedTeamId,
          onOnSelectedTeam: e.highlightTeam,
          onOnDeselectedTeam: e.unhighlightTeam,
          onOnMatchClick: e.onMatchClick,
          onOnParticipantClick: e.onParticipantClick
        }, {
          "match.title": d((i) => [
            l(e.$slots, "match.title", r(c(i)))
          ]),
          team: d((i) => [
            l(e.$slots, "team", r(c(i)))
          ]),
          "team.name": d((i) => [
            l(e.$slots, "team.name", r(c(i)))
          ]),
          "team.score": d((i) => [
            l(e.$slots, "team.score", r(c(i)))
          ]),
          feedIn: d((i) => [
            l(e.$slots, "feedIn", r(c(i)))
          ]),
          _: 3
        }, 8, ["bracket-node", "highlighted-team-id", "onOnSelectedTeam", "onOnDeselectedTeam", "onOnMatchClick", "onOnParticipantClick"])
      ])) : p("", !0),
      e.bracketNode.children[1] ? (C(), b("div", W, [
        N(o, {
          "bracket-node": e.bracketNode.children[1],
          "highlighted-team-id": e.highlightedTeamId,
          onOnSelectedTeam: e.highlightTeam,
          onOnDeselectedTeam: e.unhighlightTeam,
          onOnMatchClick: e.onMatchClick,
          onOnParticipantClick: e.onParticipantClick
        }, {
          "match.title": d((i) => [
            l(e.$slots, "match.title", r(c(i)))
          ]),
          team: d((i) => [
            l(e.$slots, "team", r(c(i)))
          ]),
          "team.name": d((i) => [
            l(e.$slots, "team.name", r(c(i)))
          ]),
          "team.score": d((i) => [
            l(e.$slots, "team.score", r(c(i)))
          ]),
          feedIn: d((i) => [
            l(e.$slots, "feedIn", r(c(i)))
          ]),
          _: 3
        }, 8, ["bracket-node", "highlighted-team-id", "onOnSelectedTeam", "onOnDeselectedTeam", "onOnMatchClick", "onOnParticipantClick"])
      ])) : p("", !0)
    ])) : p("", !0)
  ])) : p("", !0);
}
const Y = /* @__PURE__ */ y(K, [["render", X]]), Z = (e) => {
  if (!e)
    return;
  const a = e.length;
  let s = [], m = [], g = 0;
  for (let k = 0; k < a; k++) {
    if (s = e[k].matchs.map((t) => (g++, t.id = t.id ?? g, t.title = t.title ?? String(g), {
      match: t,
      children: [],
      hasParent: !!e[k + 1]
    })), m.length === 0) {
      m = s;
      continue;
    }
    for (let t = 0; t < m.length; t++) {
      const n = Math.floor(t / 2);
      s[n].children.push(m[t]);
    }
    m = s;
  }
  return s[0] || void 0;
}, _ = O({
  components: {
    BracketNode: Y
  },
  props: {
    rounds: {
      type: Array,
      default: () => []
    },
    format: {
      type: String,
      default: "default"
    },
    textColor: {
      type: String,
      default: "#ffffff"
    },
    titleColor: {
      type: String,
      default: "#9d999d"
    },
    teamBackgroundColor: {
      type: String,
      default: "#59595e"
    },
    highlightTeamBackgroundColor: {
      type: String,
      default: "#222222"
    },
    scoreBackgroundColor: {
      type: String,
      default: "#787a7f"
    },
    winnerScoreBackgroundColor: {
      type: String,
      default: "#ee7b3c"
    }
  },
  setup(e, { emit: a }) {
    const s = M(void 0), m = (h) => {
      a("onMatchClick", h);
    }, g = (h, $) => {
      a("onParticipantClick", h, $);
    }, k = (h) => {
      s.value = h;
    }, t = () => {
      s.value = void 0;
    }, n = P(() => {
      let h = e.rounds;
      return Z(h);
    }), o = P(() => ({
      "--text-color": e.textColor,
      "--title-color": e.titleColor,
      "--team-background-color": e.teamBackgroundColor,
      "--highlight-team-background-color": e.highlightTeamBackgroundColor,
      "--score-background-color": e.scoreBackgroundColor,
      "--winner-score-background-color": e.winnerScoreBackgroundColor
    }));
    return {
      highlightedTeamId: s,
      onMatchClick: m,
      onParticipantClick: g,
      highlightTeam: k,
      unhighlightTeam: t,
      recursiveBracket: n,
      cssVars: o
    };
  }
});
function x(e, a, s, m, g, k) {
  const t = v("BracketNode");
  return e.recursiveBracket ? (C(), b("div", {
    key: 0,
    class: "vt-wrapper",
    style: I(e.cssVars)
  }, [
    N(t, {
      "bracket-node": e.recursiveBracket,
      "highlighted-team-id": e.highlightedTeamId,
      onOnSelectedTeam: e.highlightTeam,
      onOnDeselectedTeam: e.unhighlightTeam,
      onOnMatchClick: e.onMatchClick,
      onOnParticipantClick: e.onParticipantClick
    }, {
      "match.title": d((n) => [
        l(e.$slots, "match.title", r(c(n)))
      ]),
      team: d((n) => [
        l(e.$slots, "team", r(c(n)))
      ]),
      "team.name": d((n) => [
        l(e.$slots, "team.name", r(c(n)))
      ]),
      "team.score": d((n) => [
        l(e.$slots, "team.score", r(c(n)))
      ]),
      feedIn: d((n) => [
        l(e.$slots, "feedIn", r(c(n)))
      ]),
      _: 3
    }, 8, ["bracket-node", "highlighted-team-id", "onOnSelectedTeam", "onOnDeselectedTeam", "onOnMatchClick", "onOnParticipantClick"])
  ], 4)) : p("", !0);
}
const te = /* @__PURE__ */ y(_, [["render", x]]);
export {
  te as TournamentBracket
};
